brut.apktool/apktool-lib/src/main/java/brut/androlib/res/decoder/StringBlock.java

```diff
@@ -54,20 +62,28 @@ public static StringBlock readWithoutChunk(ExtDataInput reader, int chunkSize) t
             block.m_styleOffsets = reader.readIntArray(styleCount);
         }
 
-        int size = ((stylesOffset == 0) ? chunkSize : stylesOffset) - stringsOffset;
-        block.m_strings = new byte[size];
-        reader.readFully(block.m_strings);
+        // #3236 - Some applications give a style offset, but have 0 styles. So probably read the string pool.
+        boolean hasStyles = stylesOffset != 0 && styleCount != 0;
+        int size = chunkSize - stringsOffset;
+
+        // If we have both strings and even just a lying style offset - lets calculate the size of the strings without
+        // accidentally parsing all the styles.
+        if (styleCount > 0) {
+            size = stylesOffset - stringsOffset;
+        }
+
+        block.m_strings = reader.readSafeByteArray(size, startPosition + chunkSize);
```

Comment:
There is something wrong here still. In a sample with an application with 0 style count, but a non-zero style offset AND more strings than possible (exceeds chunk). We are reading strings incorrectly.

Reply:
```
W: Bad string block: string entry is at 22720, past end at 22720
W: String #0 is meinstallLocation
                                 versionCode
minSdkVersiontargetSdkVersion   initOrdernameersionName
W: String #1 is 
                versionCode
minSdkVersiontargetSdkVersionrsiinitOrdernamrequiredprotectionLevel
anyDe
W: String #2 is 
W: String #3 is argetSdkVersion	initOrdernamrequiredprotectionLevel
anyDensity
normalScreensgeScreens
resizeable
          sm
W: String #4 is on	initOrdernamrequiredprotectionLevel
```

Yeah something is off still. UTF16 strings that are decoding completely wrong.

Reply:
In comparison of real strings.

```
String pool of 317 unique UTF-16 non-sorted strings, 337 entries and 0 styles using 22712 bytes:
String #0: compileSdkVersion
String #1: compileSdkVersionCodename
String #2: installLocation
String #3: versionCode
String #4: versionName
String #5: minSdkVersion
String #6: targetSdkVersion
String #7: initOrder
String #8: name
String #9: required
String #10: protectionLevel
```

Reply:
As it looks. We are reading too late and messing up the parser.

Reply:
Okay figured it out. We have 317 strings, but 337 entries. We read too many offsets and read into the string set.

Reply:
Yay.

```
I: Decoding AndroidManifest.xml with resources...
W: Bad string block: string entry is at 1300, past end at 1300
W: String #0 is compileSdkVersion
W: String #1 is compileSdkVersionCodename
W: String #2 is installLocation
W: String #3 is versionCode
W: String #4 is versionName
W: String #5 is minSdkVersion
W: String #6 is targetSdkVersion
W: String #7 is initOrder
W: String #8 is name
W: String #9 is required
W: String #10 is protectionLevel
W: String #11 is anyDensity
```

Yep just talked to myself on an issue report. For anyone reading this. I got mad at times that my research was hidden/lost, so this helps me look back in time.

My notes
```
8 - stringblock prior
16 - start of res string pool

i = 21336 - too many bytes (readSafeByteArray)

22712 - chunk size (stringblock)

22720 - at end of chunk (working?)
22804 - at end of chunk (broken)

2 - type
2 - headerSize
4 - chunkSize
4 - stringCount
4 - styleCount
4 - flags
4 - stringsOffset
4 - stylesOffset
```